File: day11/main.cpp
#include <algorithm>
#include <array>
#include <format>
#include <fstream>
#include <iostream>
#include <numeric>
#include <ranges>
#include <string>
#include <unordered_map>
#include <vector>

struct PathFinder {
    PathFinder() = delete;
    PathFinder(std::unordered_map<std::string, std::vector<std::string>> devices)
        : devices(devices) {}

    uint32_t search(std::string dest) {
        if (dest == "out") {
            return 1;
        }

        return accumulate(
            devices[dest].begin(), devices[dest].end(), 0,
            [this](uint32_t sum, std::string device) { return sum + search(device); });
    }

    uint64_t cached(std::string const& dest, std::array<std::string, 2> specials) {
        std::unordered_map<std::string, uint64_t> cache;
        std::vector<std::string> visited;
        return search2(dest, specials, visited, cache);
    }

   private:
    std::string build_key(std::string_view dest, std::vector<std::string> const& visited) {
        return std::format("{}{}", dest, std::ranges::fold_left(visited, "", std::plus{}));
    }

    uint64_t search2(std::string const& dest,
                     std::array<std::string, 2> const& specials,
                     std::vector<std::string> const& visited,
                     std::unordered_map<std::string, uint64_t>& cache) {
        std::string key = build_key(dest, visited);
        auto it = cache.find(key);
        if (it != cache.end()) {
            return it->second;
        }
        if (dest == "out") {
            bool found = std::all_of(specials.begin(), specials.end(), [&visited](std::string s) {
                return std::find(visited.begin(), visited.end(), s) != visited.end();
            });
            return found ? 1 : 0;
        }

        uint64_t sum = 0;
        for (auto device : devices[dest]) {
            auto new_visited = visited;
            if (std::find(specials.begin(), specials.end(), device) != specials.end()) {
                new_visited.push_back(device);
            }
            sum += search2(device, specials, new_visited, cache);
        }
        cache[key] = sum;
        return sum;
    }

    std::unordered_map<std::string, std::vector<std::string>> devices;
};

int main() {
    std::ifstream input("puzzle.txt");
    std::string line;
    std::unordered_map<std::string, std::vector<std::string>> devices;

    while (std::getline(input, line)) {
        auto pos = line.find(':');
        std::string name = line.substr(0, pos);
        std::string rest = line.substr(pos + 2);
        std::vector<std::string> dests =
            rest | std::views::split(' ') | std::ranges::to<std::vector<std::string>>();
        devices[name] = dests;
    }

    PathFinder finder(devices);
    std::cout << finder.search("you") << std::endl;

    std::array<std::string, 2> specials = {"dac", "fft"};
    std::cout << finder.cached("svr", specials) << std::endl;
    return 0;
}

File: day11/src/main.rs
use std::collections::HashMap;

struct PathFinder {
    devices: HashMap<String, Vec<String>>,
}

impl PathFinder {
    fn new(devices: HashMap<String, Vec<String>>) -> PathFinder {
        PathFinder { devices }
    }

    fn search(&self, dest: String) -> u32 {
        if dest == "out" {
            return 1;
        }

        self.devices
            .get(&dest)
            .unwrap()
            .iter()
            .map(|device| self.search(device.to_string()))
            .sum()
    }

    fn cached(&self, dest: &str, specials: &[&str]) -> u64 {
        let mut cache: HashMap<String, u64> = HashMap::new();
        let visited: Vec<String> = Vec::new();
        self.search2(dest, specials, visited, &mut cache)
    }

    fn build_key(dest: &str, visited: &[String]) -> String {
        format!(
            "{}{}",
            dest,
            visited.iter().map(|v| v.to_string()).collect::<String>()
        )
    }

    fn search2(
        &self,
        dest: &str,
        specials: &[&str],
        visited: Vec<String>,
        cache: &mut HashMap<String, u64>,
    ) -> u64 {
        let key = PathFinder::build_key(dest, &visited);
        if let Some(value) = cache.get(&key) {
            return *value;
        }
        if dest == "out" {
            if specials
                .iter()
                .all(|device| visited.contains(&device.to_string()))
            {
                return 1;
            }
            return 0;
        }

        let sum = self
            .devices
            .get(dest)
            .unwrap()
            .iter()
            .map(|device| {
                let mut new_visited = visited.clone();
                if specials.contains(&(device.as_str())) {
                    new_visited.push(device.to_string());
                }
                self.search2(device, specials, new_visited, cache)
            })
            .sum();

        cache.insert(key, sum);
        sum
    }
}

fn main() {
    let file = include_str!("../puzzle.txt");
    let devices: HashMap<String, Vec<String>> = file
        .lines()
        .map(|line| {
            let temp = line.replace(":", "");
            let mut parts = temp.split(" ");
            let key = parts.next().unwrap().to_string();
            let values = parts.map(|v| v.to_string()).collect();
            (key, values)
        })
        .collect();

    let finder = PathFinder::new(devices);
    println!("{}", finder.search("you".to_string()));

    let specials = ["dac", "fft"];
    println!("{}", finder.cached("svr", &specials));
}
